### PR TITLE
Replace TypedDict with dataclasses

### LINKED ISSUE #487: Expose an API to make it easy to parse SSLyze's JSON output
**Is your feature request related to a problem? Please describe.**
We have a tool parsing SSLyze JSON output and giving us information about weaknesses. (Sidenote: we really like SSLyze for not doing this, as we think, interpretation is a "personal" decision).

Right now we read and parse the JSON and then handle it with multiple dict-accesses (``json["scan_command_results"]`` or ``json.get("some_plugin")`` or ``if "some_plugin" in json:``). This leads to regular errors and mistakes in our code base (either we try to access a key which is not there, as it might be optional, or we access a field by ``.get`` not realizing a typo and always get ``False``). So we thought about how to improve and a natural way would be using a library like https://github.com/konradhalas/dacite to read the entire JSON structure back in the corresponding dataclasses. By this, we could write save and type-checked code afterwards and would easily be warned about access to optional or misspelled fields (by ``mypy`` or ``flake8``).

I tried one example JSON yesterday and one big issue is the use of ``TypedDict``s for ``ScanCommandResultsDict`` and ``ScanCommandExtraArgumentsDict``:
https://github.com/nabla-c0d3/sslyze/blob/20d6fb6c4118ff54d67bb3d176c99136bc09708b/sslyze/scanner.py#L84
https://github.com/nabla-c0d3/sslyze/blob/20d6fb6c4118ff54d67bb3d176c99136bc09708b/sslyze/scanner.py#L57
I just patched quick and dirty ``TypedDict`` handling in https://github.com/konradhalas/dacite and got it read at least the example I tested fully back to the ``_SslyzeOutputAsJson`` class. Very nice!

**Describe the solution you'd like**
I think, ``TypedDict`` handling for a library like https://github.com/konradhalas/dacite might not be really necessary and thus propose to ban ``TypedDict``s in SSLyze and use dataclasses for the two mentioned containers like for all other container-classes in SSLyze.

**Describe alternatives you've considered**
Patching https://github.com/konradhalas/dacite as mentioned before.

**Additional context**
As on a first glance it looks like not a very big deal to me, I would be happy to create a PR for this in the upcoming weeks. But as I have a working "quick and dirty" solution right now, I first wanted to get in touch and ask, if there is any reason, why these classes are still ``TypedDict``s, and if you would consider a PR to use dataclasses.

I would be happy about feedback and your thoughts on this @nabla-c0d3.

BR

### PR DESCRIPTION
**Pull request by dquitmann-op**
Fix #487: Replace TypedDicts with dataclasses

### CODE REVIEW

**Comment by dquitmann-op** [comment]
First try of replacing TypedDict ``ScanCommandResultsDict`` with dataclass ``ScanCommandResults``. As this is a first PoC to discuss, the tests are not yet all correctly implemented.

Also similar changes would have to be made for ``ScanCommandExtraArgumentsDict``.

Would like to get feedback about the approach.

File: sslyze/scanner.py [hunk 1]
```diff
@@ -264,7 +277,7 @@ def get_results(self) -> Iterable[ServerScanResult]:
                         result = implementation_cls.result_for_completed_scan_jobs(
                             server_info, list(completed_scan_jobs)
                         )
-                        server_scan_results[scan_cmd] = result
+                        server_scan_results = dataclasses_replace(server_scan_results, scan_cmd=result)
```
Comment (dquitmann-op) [inline sslyze/scanner.py]: As I wanted to make the dataclass ``ScanCommandResults`` ``frozen`` as all other, we have to use ``replace`` and create a new Instance on every new completed scan job. As this is only a small container of other objects (which do not have to be copied/transferred), this should be ok (in performance).

The alternative would be to create a "unfrozen" version and then somehow "freeze" it.

File: tests/factories.py [hunk 2]
```diff
@@ -114,13 +114,13 @@ class ServerScanResultFactory:
...
-        scan_commands_results: Optional[ScanCommandResultsDict] = None,
+        scan_commands_results: Optional[ScanCommandResults] = None,
         scan_commands_errors: Optional[ScanCommandErrorsDict] = None,
     ) -> ServerScanResult:
-        final_results: ScanCommandResultsDict = (
+        final_results: ScanCommandResults = (
             scan_commands_results
             if scan_commands_results
-            else {ScanCommand.TLS_COMPRESSION: CompressionScanResult(supports_compression=True)}
+            else ScanCommandResults(tls_compression=CompressionScanResult(supports_compression=True))
```
Comment (dquitmann-op) [inline tests/factories.py:123]: I did not see any way of using the constant as key. Any suggestion would be welcome.

File: sslyze/scanner.py [hunk 3]
```diff
@@ -80,30 +80,43 @@ def __post_init__(self) -> None:
...
+            key = field_.name
+            if getattr(self, key) is not None:
+                yield key
+
+    def items(self) -> Iterator[Tuple[str, Any]]:
+        for key in self.keys():
+            yield key, getattr(self, key)
+
+    def __len__(self) -> int:
+        return len(list(self.keys()))
```
Comment (dquitmann-op) [inline sslyze/scanner.py:119]: The ``keys``, ``items`` and ``__len__`` functions are mainly for ease of use with the current implementation.

**Comment by dquitmann-op** [comment]
A noteable side-effect of the current implementation: as the dataclass ``ScanCommandResults`` has default ``None`` values for all field, these are all exported every time with ``--json_out``, e.g. `` sslyze --robot google.de --json_out test.json`` gives something like:

```json
...
"scan_commands_results": {
    "certificate_info": null,
    "elliptic_curves": null,
    "heartbleed": null,
    "http_headers": null,
    "openssl_ccs_injection": null,
    "robot": {
        "robot_result": "NOT_VULNERABLE_NO_ORACLE"
    },
    "session_renegotiation": null,
    "session_resumption": null,
    "session_resumption_rate": null,
    "ssl_2_0_cipher_suites": null,
    "ssl_3_0_cipher_suites": null,
    "tls_1_0_cipher_suites": null,
    "tls_1_1_cipher_suites": null,
    "tls_1_2_cipher_suites": null,
    "tls_1_3_cipher_suites": null,
    "tls_1_3_early_data": null,
    "tls_compression": null,
    "tls_fallback_scsv": null
},
...
```

If this is strictly unwanted, one could filter the output in the JSON-generating code. But I do not really see any problem/issue (other than wasting some bytes).

**Comment by dquitmann-op** [comment]
``pipenv run invoke test`` still fails, as The mocked ScanCommands are not saved in ``ScanCommandResults`` (as there are no fields for them). I'm not 100% sure about this Mocking and how to solve this most elegant (probably one wants to also mock the ``ScanCommandResults`` dataclass to also include fields for the mocked commands?!)

File: sslyze/scanner.py (hunk 3 above)
Comment (nabla-c0d3) [reply sslyze/scanner.py:119]: I thought about it and I think I will just release this change as part of the next major release (5.0.0) so you don't need to maintain backward compatibility. You can remove these functions.

File: tests/factories.py (hunk 2 above)
Comment (nabla-c0d3) [reply tests/factories.py:123]: This looks good as it is.

File: tests/web_servers/scan_localhost.py [hunk 4]
```diff
@@ -80,15 +80,15 @@ def main() -> None:
             ScanCommand.CERTIFICATE_INFO,
             ScanCommand.TLS_COMPRESSION,
         }
-        if server_scan_result.scan_commands_results.keys() != expected_scan_command_results:
+        if set(server_scan_result.scan_commands_results.keys()) != expected_scan_command_results:
```
Comment (nabla-c0d3) [inline tests/web_servers/scan_localhost.py:135]: After removing keys() this will need to be updated.

File: sslyze/scanner.py [hunk 5]
```diff
@@ -264,7 +277,8 @@ def get_results(self) -> Iterable[ServerScanResult]:
                         result = implementation_cls.result_for_completed_scan_jobs(
                             server_info, list(completed_scan_jobs)
                         )
-                        server_scan_results[scan_cmd] = result
+                        scan_cmd_str = cast(str, scan_cmd)
+                        server_scan_results = dataclasses_replace(server_scan_results, **{scan_cmd_str: result})
```
Comment (nabla-c0d3) [inline sslyze/scanner.py]: In this case let's just not make ScanCommandResults frozen - I think that's fine.

File: sslyze/scanner.py (hunk 3 above)
Comment (nabla-c0d3) [reply sslyze/scanner.py:119]: Also you don't need to rely on getattr() to fetch a result; you can just access the field directly.

**Review by nabla-c0d3 — COMMENTED** [review]
I had a first look and left a few comments; it's going into the right direction. Thanks for your contribution!

**Comment by nabla-c0d3** [comment]
> If this is strictly unwanted, one could filter the output in the JSON-generating code. But I do not really see any problem/issue (other than wasting some bytes).

I think it's fine with null fields.

File: sslyze/scanner.py (hunk 5 above)
Comment (dquitmann-op) [reply sslyze/scanner.py]: Did this right now.

File: sslyze/scanner.py (hunk 3 above)
Comment (dquitmann-op) [reply sslyze/scanner.py:119]: I also thought about it again. And clearly there is two "variants of access" for this dataclass:
1. direct access to one field (can be done by ``server_scan_results.field_name`` or ``getattr(server_scan_results, field_name_var)``
1. iterating over/counting/using all fields (sometimes only the keys).

As we only want to count/iterate over the fields which are not ``None`` in the second case, some convenience wrapper/functions are needed imho (as you do not want to implement this logic multiple times). One can this/discuss about the names (which are very dict-like right now; I happily change them, if you have any preferences), but I think we need them, as this dataclass should be somehow "iterable".

> Also you don't need to rely on getattr() to fetch a result; you can just access the field directly.

I'm not aware, how to access a dataclass field whose name I have in a variable without ``getattr()``?! Please review the code again and give me hint how to do this if its possible.

**Comment by dquitmann-op** [comment]
I rebased the PR onto the newest release and unfreezed the dataclass.

Still todo:
* discuss exact implementation details (see above discussions)
* move ``ScanCommandExtraArgumentsDict`` to a dataclass
* get tests working (see https://github.com/nabla-c0d3/sslyze/pull/493#issuecomment-775261703)

**Comment by nabla-c0d3** [comment]
I've taken a slight different approach, described in https://github.com/nabla-c0d3/sslyze/issues/487.

**Pull request closed without merging**